hack/generator/pkg/astmodel/conversion_function_builder.go

```diff
@@ -0,0 +1,494 @@
+
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astbuilder"
+)
+
+// ConversionParameters are parameters for converting between a source type and a destination type.
+type ConversionParameters struct {
+	Source            dst.Expr
+	Destination       dst.Expr
+	SourceType        Type
+	DestinationType   Type
+	NameHint          string
+	ConversionContext []Type
+	AssignmentHandler func(destination, source dst.Expr) dst.Stmt
+}
+
+// ClonedSource clones the Source field.
+func (params ConversionParameters) ClonedSource() dst.Expr {
+	return dst.Clone(params.Source).(dst.Expr)
+}
+
+// ClonedDestination clones the Destination field.
+func (params ConversionParameters) ClonedDestination() dst.Expr {
+	return dst.Clone(params.Destination).(dst.Expr)
+}
+
+// WithSource returns a new ConversionParameters with the updated Source.
+func (params ConversionParameters) WithSource(source dst.Expr) ConversionParameters {
+	result := params.copy()
+	result.Source = source
+
+	return result
+}
+
+// WithSourceType returns a new ConversionParameters with the updated SourceType.
+func (params ConversionParameters) WithSourceType(t Type) ConversionParameters {
+	result := params.copy()
+	result.SourceType = t
+
+	return result
+}
+
+// WithDestination returns a new ConversionParameters with the updated Destination.
+func (params ConversionParameters) WithDestination(destination dst.Expr) ConversionParameters {
+	result := params.copy()
+	result.Destination = destination
+
+	return result
+}
+
+// WithDestinationType returns a new ConversionParameters with the updated DestinationType.
+func (params ConversionParameters) WithDestinationType(t Type) ConversionParameters {
+	result := params.copy()
+	result.DestinationType = t
+
+	return result
+}
+
+// WithAssignmentHandler returns a new ConversionParameters with the updated AssignmentHandler.
+func (params ConversionParameters) WithAssignmentHandler(
+	assignmentHandler func(result dst.Expr, destination dst.Expr) dst.Stmt) ConversionParameters {
+	result := params.copy()
+	result.AssignmentHandler = assignmentHandler
+
+	return result
+}
+
+// AssignmentHandlerOrDefault returns the AssignmentHandler or a default assignment handler if AssignmentHandler was nil.
+func (params ConversionParameters) AssignmentHandlerOrDefault() func(destination, source dst.Expr) dst.Stmt {
+	if params.AssignmentHandler == nil {
+		return AssignmentHandlerAssign
+	}
+	return params.AssignmentHandler
+}
+
+// TODO: Consider a better way to do this, using KnownLocalsSet?
+// CountArraysAndMapsInConversionContext returns the number of arrays/maps which are in the conversion context.
+// This is to aid in situations where there are deeply nested conversions (i.e. array of map of maps). In these contexts,
+// just using a simple assignment such as "elem := ..." isn't sufficient because elem my already have been defined above by
+// an enclosing map/array conversion context. We use the depth to do "elem1 := ..." or "elem7 := ...".
```

Comment:
For the storage conversions I reduced the number of collisions by using distinct suffixes - `Key` and `Value` for maps, `Item` and `Index` (if I recall correctly) for lists, and so on.

Reply:
I'd like to make some improvements to the naming, but I think I should leave that part for a follow up PR, as any changes to how we generate names in the code is going to explode the PR side in the generated files - thoughts?

Reply:
Followup PR is fine.